nixos/modules/services/backup/bupstash.nix

```diff
@@ -0,0 +1,666 @@
+                  Only create the entry if the command exited with a successful status code.
+                '';
+              };
+
+              key = mkOption {
+                type = nullOr path;
+                default = null;
+                description = ''
+                  Path to the key that will be used to encrypt the data.
+                  See [the documentation](https://bupstash.io/doc/guides/Secure%20Offline%20Keys.html) to securely use keys.
+                '';
+              };
+
+              keyCommand = mkOption {
+                type = nullOr str;
+                default = null;
+                description = ''
+                  Run a command that prints the key to stdout. Mutually exclusive with the `key` option.
+                  Useful to retrieve the key from password managers.
+                '';
+              };
+
+              user = mkOption {
+                type = str;
+                default = defaultClientUser;
+                description = ''
+                  User that executes the backup job. Will run as a [dynamic user](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=) by default.
+                  If the backup needs to run under a static UID, declaring a user account with the same name is sufficient.
+                '';
+              };
+
+              group = mkOption {
+                type = nullOr str;
+                default = null;
+                description = "Group for the backup job. When no group is provided (the default behaviour).";
+              };
+
+              preHook = mkOption {
+                type = lines;
+                default = "";
+                description = ''
+                  Commands that are run before the put job is run. Executed with root permissions.
+                  Useful to create snapshots and perform other preparations.
+                '';
+                example = literalExpression ''
+                  # https://openzfs.github.io/openzfs-docs/man/8/zfs-snapshot.8.html
+                  ''${config.boot.zfs.package}/bin/zfs snapshot -r pool/state@backup
+                '';
+              };
+
+              postHook = mkOption {
+                type = lines;
+                default = "";
+                description = ''
+                  Commands that are run before the put job is run.
+                  Useful to remove snapshots and perform other cleanup tasks.
+                '';
+                example = literalExpression ''
+                  # https://openzfs.github.io/openzfs-docs/man/8/zfs-snapshot.8.html
+                  ''${config.boot.zfs.package}/bin/zfs destroy -r pool/state@backup
+                '';
+              };
+
+              repositories = mkOption {
+                type = attrsOf (
+                  submodule (
+                    { config, ... }:
+                    {
+                      options = {
```

Comment:
I'm sorry for the back and forth, but this section will require a bit of rework. My proposal:

- The `ssh` kind has a low value/complexity ratio in my eyes. I'd rather like to not have it and tell users in the documentation to use `ssh` commands or `ssh://` URIs.
- Then the `kind` option can be removed, by simply using `uri` and `command` exclusively and default null values. (Agreed that this is not great code wise, but I think better than exposing a `type` option to the Nix users, as this is not really that common in NixOS.)
- Not sure about the name of `uri` and `command`, although I see that `path` instead of `uri` would be equally confusing. Maybe it will suffice to document that these map to the `BUPSTASH_REPOSITORY` and `BUPSTASH_REPOSITORY_COMMAND` env variables.
- I'm not sure how important per-repo CLI flags are and would propose to remove `enableSendLog`.
- Maybe rename `writePaths` into `writablePaths` or something like that. Also, what's the use-case for specifying any paths beyond the repository path? Maybe this option could be removed as well.

Reply:
When you want to create a local repository automatically, there is nothing for systemd to bind to, so we need to set `writablePaths` to the parent directory of where the repo will be

Reply:
To remove the `writablePaths` option, we could explicitely disallow setting `createRepository` to true when the repository is a local one, and telling the user to create it first (as it is a local one, running `bupstash-init` is quite trivial), what do you think @mweinelt @piegamesde ?

Reply:
I'm torn on the `writablePaths` idea, given that issue. This mostly affects backups to local storage, for which I would highly recommend going with explicit user/group settings, because otherwise uid/gid won't be stable, and reasonable permissions on the parent directory.

Please read through the other points that were made and let's discuss the details.

> Not sure about the name of uri and command, although I see that path instead of uri would be equally confusing. Maybe it will suffice to document that these map to the BUPSTASH_REPOSITORY and BUPSTASH_REPOSITORY_COMMAND env variables.

i think this is one of the more important points, and going back to something simpler is the solution:

```nix
# BUPSTASH_REPOSITORY
repositories = mkOption {
  type = listOf str;
  default = [];
};

repositoryCommands = mkOption {
  type = listOf str;
  default = [];
}
```

This makes the fewest assumptions and offers the highest customizability, and requires no further escape hatches.

Reply:
I think the highest customizability would be `attrsOf str`, if not, we would still need to find a way to generate a readable name from a command or a path

Reply:
@mweinelt @piegamesde,
To recap the latest push,
- there is now an attrset `repositories` mapped to `BUPSTASH_REPOSITORY`s and `repositoryCommands` mapped to `BUPSTASH_REPOSITORY_COMMAND`
- I removed the `BindPaths` thing
- The option `createRepositories` is a list of names that must defined in `repositories` or `repositoryCommand`
- The `enableSendlog` option has been removed